File: C/binarysearch/rosetta/rosetta.c
#include <stdio.h>

int bsearch (int *a, int n, int x) {
    int i = 0, j = n - 1;
    while (i <= j) {
        int k = i + ((j - i) / 2);
        if (a[k] == x) {
            return k;
        }
        else if (a[k] < x) {
            i = k + 1;
        }
        else {
            j = k - 1;
        }
    }
    return -1;
}

int bsearch_r (int *a, int x, int i, int j) {
    if (j < i) {
        return -1;
    }
    int k = i + ((j - i) / 2);
    if (a[k] == x) {
        return k;
    }
    else if (a[k] < x) {
        return bsearch_r(a, x, k + 1, j);
    }
    else {
        return bsearch_r(a, x, i, k - 1);
    }
}

int main() {
    int n = 1000;
    int a[n];
    srand(time(NULL));

    for (int i = 0; i < n; i++) {
        a[i] = rand() % 10000;  // Gera valores aleatórios entre 0 e 9999
    }

    // Realiza buscas (sem prints)
    int x = rand() % 10000;  // Valor aleatório para busca
    bsearch(a, n, x);

    x = rand() % 10000;  // Outro valor aleatório para busca
    bsearch_r(a, x, 0, n - 1);

    return 0;
}

File: Java/binarysearch/Rosetta/Rosetta.java
import java.util.Arrays;

public class Rosetta {

    public static int binarySearch(int[] nums, int check) {
        int hi = nums.length - 1;
        int lo = 0;
        while (hi >= lo) {
            int guess = (lo + hi) >>> 1;  // from OpenJDK
            if (nums[guess] > check) {
                hi = guess - 1;
            } else if (nums[guess] < check) {
                lo = guess + 1;
            } else {
                return guess;
            }
        }
        return -1;
    }

    public static void main(String[] args) {
        int[] haystack = new int[1000];
        
        // Preencher o array com valores aleatórios entre 100 e 1000
        for (int i = 0; i < haystack.length; i++) {
            haystack[i] = (int) (Math.random() * 901) + 100; // Valores aleatórios entre 100 e 1000
        }
        
        // Ordenar o array para que a busca binária funcione corretamente
        Arrays.sort(haystack);
        
        // Exemplo de valor a ser procurado (gerado aleatoriamente ou definido)
        int needle = (int) (Math.random() * 901) + 100; // Valor aleatório entre 100 e 1000
        
        // Realizar a busca binária
        binarySearch(haystack, needle);
    }
}
